engines/sci/graphics/cursor32.cpp

```diff
@@ -0,0 +1,394 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "common/rational.h"        // for Rational, operator*
+#include "common/system.h"          // for OSystem, g_system
+#include "graphics/cursorman.h"     // for CursorMan
+#include "sci/graphics/celobj32.h"  // for CelObjView, CelInfo32, Ratio
+#include "sci/graphics/cursor32.h"
+#include "sci/graphics/frameout.h"  // for GfxFrameout
+
+namespace Sci {
+
+GfxCursor32::GfxCursor32() :
+	_hideCount(0),
+	_position(0, 0),
+	_writeToVMAP(false) {
+	CursorMan.showMouse(false);
+}
+
+void GfxCursor32::init(const Buffer &vmap) {
+	_vmap = vmap;
+	_vmapRegion.rect = Common::Rect(_vmap.screenWidth, _vmap.screenHeight);
+	_vmapRegion.data = (byte *)_vmap.getPixels();
+	_restrictedArea = _vmapRegion.rect;
+}
+
+GfxCursor32::~GfxCursor32() {
+	CursorMan.showMouse(true);
+	free(_cursor.data);
+	free(_cursorBack.data);
+	free(_drawBuff1.data);
+	free(_drawBuff2.data);
+	free(_savedVmapRegion.data);
+}
+
+void GfxCursor32::hide() {
+	if (_hideCount++) {
+		return;
+	}
+
+	if (!_cursorBack.rect.isEmpty()) {
+		drawToHardware(_cursorBack);
+	}
+}
+
+void GfxCursor32::revealCursor() {
+	_cursorBack.rect = _cursor.rect;
+	_cursorBack.rect.clip(_vmapRegion.rect);
+	if (_cursorBack.rect.isEmpty()) {
+		return;
+	}
+
+	readVideo(_cursorBack);
+	_drawBuff1.rect = _cursor.rect;
+	copy(_drawBuff1, _cursorBack);
+	paint(_drawBuff1, _cursor);
+	drawToHardware(_drawBuff1);
+}
```

Comment:
Why can't we use SCUMMVM system functions for the cursor just like in the other cursor.cpp?

I'm not sure if that's the case but platforms could implement cursors in a way that performs better than doing all of that by ourselves.